lib/Conversion/ExportVerilog/PrepareForEmission.cpp

```diff
@@ -764,6 +764,40 @@ static LogicalResult legalizeHWModule(Block &block,
       continue;
     }
 
+    // Force array index expressions to be a simple name when the
+    // `mitigateVivadoArrayIndexConstPropBug` option is set.
+    if (options.mitigateVivadoArrayIndexConstPropBug &&
+        isa<ArraySliceOp, ArrayGetOp, ArrayIndexInOutOp,
+            IndexedPartSelectInOutOp, IndexedPartSelectOp>(&op)) {
+
+      // Check if the index expression is already a wire.
+      WireOp wireOp;
+      ReadInOutOp readOp;
+      if (auto maybeReadOp =
+              op.getOperand(1).getDefiningOp<sv::ReadInOutOp>()) {
+        if (auto maybeWireOp = maybeReadOp.getInput().getDefiningOp<WireOp>()) {
+          wireOp = maybeWireOp;
+          readOp = maybeReadOp;
+        }
+      }
+
+      // Insert a wire and read if necessary.
+      ImplicitLocOpBuilder builder(op.getLoc(), &op);
+      if (!wireOp) {
+        wireOp =
+            builder.create<WireOp>(op.getOperand(1).getType(), "_ARRAY_IDX");
```

Comment:
I think the op could be in a procedural region. In that case I think we have to use logic+bpassign. 
```
hw.module @Foo(%a: !hw.array<16xi1>, %b : i4) -> () {
  %c1_i4 = hw.constant 1 : i4
  %0 = comb.add %b, %c1_i4 : i4
  sv.initial {
    %1 = hw.array_get %a[%0] : !hw.array<16xi1>, i4
    sv.verbatim "{{0}}" (%1): i1
  }
  hw.output
}
```

Reply:
Great point! Added a fix for that :+1: